// ng-gapi/esm2015/GoogleApiService.js
import { __decorate, __param } from "tslib";
import { Inject, Injectable, InjectionToken } from '@angular/core';
import { GoogleApiConfig } from './config/GoogleApiConfig';
import { Observable } from 'rxjs';
import * as ɵngcc0 from '@angular/core';
export let NG_GAPI_CONFIG = new InjectionToken('ng-gapi.config');
let GoogleApiService = class GoogleApiService {
    constructor(config) {
        this.gapiUrl = 'https://apis.google.com/js/api.js';
        this.observers = [];
        this.config = new GoogleApiConfig(config);
        this.loadGapi().subscribe();
    }
    onLoad() {
        return this.loadGapi();
    }
    getConfig() {
        return this.config;
    }
    loadGapi() {
        return new Observable((observer) => {
            if (this.gapiLoaded) {
                observer.next(true);
                observer.complete();
            }
            else if (!this.node) {
                /**
                 * script element has not yet been added to document
                 */
                this.node = document.createElement('script');
                this.node.async = true;
                this.node.src = this.gapiUrl;
                this.node.type = 'text/javascript';
                this.node.onload = () => {
                    this.gapiLoaded = true;
                    while (this.observers.length) {
                        const observer = this.observers.shift();
                        observer.next(true);
                        observer.complete();
                    }
                    this.node = undefined;
                };
                this.node.onerror = () => {
                    this.node = undefined;
                };
                document.getElementsByTagName('head')[0].appendChild(this.node);
            }
            else {
                /**
                 * script is in the middle of being loaded
                 */
                this.observers.push(observer);
            }
        });
    }
};
GoogleApiService.ɵfac = function GoogleApiService_Factory(t) { return new (t || GoogleApiService)(ɵngcc0.ɵɵinject(NG_GAPI_CONFIG)); };
GoogleApiService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: GoogleApiService, factory: function (t) { return GoogleApiService.ɵfac(t); } });
GoogleApiService = __decorate([ __param(0, Inject(NG_GAPI_CONFIG))
], GoogleApiService);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(GoogleApiService, [{
        type: Injectable
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [NG_GAPI_CONFIG]
            }] }]; }, null); })();
export { GoogleApiService };
