// ng-gapi/esm2015/GoogleAuthService.js
import { __decorate, __param } from "tslib";
import { Injectable, Inject } from '@angular/core';
import { GoogleApiService } from './GoogleApiService';
import { Observable, of } from 'rxjs';
import { mergeMap } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
let GoogleAuthService = class GoogleAuthService {
    constructor(googleApi) {
        this.googleApi = googleApi;
        this.GoogleAuth = undefined;
        this.googleApi.onLoad().subscribe(() => {
            this.loadGapiAuth().subscribe();
        });
    }
    getAuth(newInstance = false) {
        if (!this.GoogleAuth || newInstance) {
            return this.googleApi.onLoad()
                .pipe(mergeMap(() => this.loadGapiAuth()));
        }
        return of(this.GoogleAuth);
    }
    loadGapiAuth() {
        return new Observable((observer) => {
            gapi.load('auth2', () => {
                gapi.auth2.init(this.googleApi.getConfig().getClientConfig()).then((auth) => {
                    this.GoogleAuth = auth;
                    observer.next(auth);
                    observer.complete();
                }).catch((err) => observer.error(err));
            });
        });
    }
};
GoogleAuthService.ɵfac = function GoogleAuthService_Factory(t) { return new (t || GoogleAuthService)(ɵngcc0.ɵɵinject(GoogleApiService)); };
GoogleAuthService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: GoogleAuthService, factory: function (t) { return GoogleAuthService.ɵfac(t); } });
GoogleAuthService = __decorate([ __param(0, Inject(GoogleApiService))
], GoogleAuthService);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(GoogleAuthService, [{
        type: Injectable
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [GoogleApiService]
            }] }]; }, null); })();
export { GoogleAuthService };
